[decisions] constructor functions (#4244)
ElektraInitiative/libelektra: doc/decisions/constructor_functions.md

# Constructor Functions

## Problem

The constructor functions for creating Keys and KeySets must be easy to be invoked from other languages without complicated workarounds.

> In particular, another implementation should not need to resort to adding a small amount of C to "bridge the gap".

Reviewer: In which binding do we have C code only for constructors? In those we do not have any C code (like JNA), keyNew and ksNew works without any bridge code. In those we already have C code, fixing only the constructors seems pointless to me.

@lawli3t What is the situation in Rust?

Author: The Go binding definitely has some wrapper C code already. 

https://github.com/ElektraInitiative/go-elektra/blob/0462a716b69709257ad2e7b0b881eef645a0160f/kdb/key.go#L3-L13

The existing Rust binding doesn't have wrapper code, but it tries it's best to ignore the vararg part of the C API:

https://github.com/ElektraInitiative/libelektra/blob/eac1c6a74cb75bb84ebfae2247b3cc531259ded8/src/bindings/rust/elektra/src/writeable.rs#L55-L75

The Java binding does use the varargs, but needs some pretty horrible translation code:

https://github.com/ElektraInitiative/libelektra/blob/eac1c6a74cb75bb84ebfae2247b3cc531259ded8/src/bindings/jna/libelektra/src/main/java/org/libelektra/Key.java#L116-L137

which at the end just calls another layer of translation code: 

https://github.com/ElektraInitiative/libelektra/blob/eac1c6a74cb75bb84ebfae2247b3cc531259ded8/src/bindings/jna/libelektra/src/main/java/org/libelektra/Key.java#L92-L102

The SWIG bindings are based on C++ and AFAICT the SWIG bindings and C++ are the only ones where the varargs API from C can actually be used.

In summary, maybe the phrasing about "C to bridge the gap" was not quite accurate. It applies in general IMO that there shouldn't be a need for bridge code, but it is not the current problem. Nonetheless, I'd say creating a varargs binding in the core API which then can't always be used directly is a bad idea. That is why I would avoid varargs in the core API and either completely avoid them or at the very least put them into a separate lowlevel C library. It simply allows more flexibility and allows the bindings to better fit their languages.

AFAIK the main reason for the whole varargs API is that it allows creating a whole `KeySet` in a single expression. In C this can only be achieved, if a `KeySet` is created by single function that receives all the `Key`s _and_ a `Key` can also be constructed with a single function. But in other languages there are different, more idomatic possibilities. In Java, Rust and many other languages, a builder pattern makes more sense. I don't see the point in burying an impractical varargs API underneath those builders. Especially not if the varargs API still just calls `keySetName`, `keySetMeta`, etc.

> fixing only the constructors seems pointless to me

To me it seems pointless to have varargs functions. I don't actually know of a single argument in favour of varargs other than single expression thing (which I've now shown multiple times can be achieved differently too). IMO varargs just remove more of the already very lax type checks from the C compiler.

---

At the end there is one thing I am 100% certain of. There is absolutely no place for varargs APIs in the `libelektra-core` API. We can argue about whether or not we want to use them in `libelektra-lowlevel-c` (which targets only C), but for the core API I don't see any reason to have a varargs API.

Reviewer: > At the end there is one thing I am 100% certain of. There is absolutely no place for varargs APIs in the libelektra-core API. We can argue about whether or not we want to use them in libelektra-lowlevel-c (which targets only C), but for the core API I don't see any reason to have a varargs API.

If there is no massive performance benefit (with a proper keyNew allocating everything at once), I agree. If there is a massive performance benefit, the status quo makes sense: bindings wrap their keyNew in a way it is suitable for them . E.g. if they have a constructor that always creates key+value, they would call e.g. `keyNew(key, KEY_VALUE, value, KEY_END)`.

So we need a benchmark (that shows that allocations are not the bottleneck of keyNew) to make a proper decision.

Btw. I am very thankful that you are so persisting in these things as this is the only way to look deeply enough to make really good decisions.

# Constructor Functions

## Problem

> The constructor functions that `libelektra-core` defines for creating Keys and KeySets must be possible to implement in other languages without complicated workarounds.

Reviewer: Afaik this problem does not exist, it was possible for every language we support?

```suggestion
The constructor functions for creating Keys and KeySets must be easy to be invoked from other languages without complicated workarounds.
```

In particular, another implementation should not need to resort to adding a small amount of C to "bridge the gap".

## Constraints

- Constructor functions should use as little resources as feasible.
  In particular, the number of `malloc`s and `memcpy` should be kept in check.
  Temporary allocations during key creation should be avoided.
- Using the escaped name in the constructor function should be avoided, since the unescaping process can be expensive.

## Assumptions

- The constructor functions from `libelektra-core` will not or only very rarely be used directly by users.
- For all languages supported by Elektra (including C), there is secondary library on top of `libelektra-core` that provides more idiomatic APIs.

> - Creating a key with a name, no value and no metadata and then calling `keySetValue`/`keySetMeta` does not create significant overhead compared to creating the key with value and metadata directly.

Reviewer: This is the key of the whole decision. I don't think we should base this on assumptions but actually benchmark keyNew if avoiding calling `keySetValue`/`keySetMeta` is a relevant topic or not.

Simply leaving it as is is also totally possible. In any case it is good that we now have this decision.

Author: To clarify, the assumption says:

There is no significant overhead when switching from
```c
Key * k = keyNew ("/foo", KEY_VALUE, "value", KEY_META, "meta:/type", "string", KEY_META, "meta:/bar", "baz", KEY_END);
```

to

```c
Key * k =keyNew ("/foo", KEY_END);
keySetString (k, "value");
keySetMeta (k, "meta:/type", "string");
keySetMeta (k, "meta:/bar", "baz");
```

This is currently true, because `keyNew` internally just call `keySetString`/`keySetMeta`. (*) The assumption is just that, if `keyNew` were to directly set the value and directly construct the correct `struct _KeySet`, there still wouldn't be a significant difference. Since there are no optimizations that `keyNew` could do, that couldn't be done by `keySetValue`/`keySetMeta`, I think this is a fair assumption.

Additionally, the assumption is actually only semi-relevant to the decision. In fact, the actual decision, makes the assumption less relevant. Since `keyCreate` could still use the private API and directly construct the `struct _Key`, any optimization could simply be implemented there.

---

(*) In theory, there could be (significant) difference between 1 call and 4 calls to dynamically linked symbols. However, if that change is significant, the whole idea of a minimal API becomes question able. If the API is minimal, you always need more calls to do complicated stuff, compared to an API that has functions for everything.

 * If `nameSize == 0`, the returned key is the root key of @p ns and @p name is ignored.
 * Otherwise, the name of the key within the namespace will be @p name of length @p nameSize (including null terminator).
 * Therefore, the returned key `k` fullfills:
 * `keyNameSize (k) == nameSize + 2`
 * (one byte for the namespace, and one for the separator between namespace and the rest of the name)
 *
 *
 * The key will have no value and no metadata.
 */
Key * elektraKeyNew (elektraNamespace ns, const char * name, size_t nameSize);

/**
 * Allocates a new keyset with space reserved for at least @p keyCount keys.
 * If `keyCount == 0`, the keyset will be empty (with no space for keys reserved) and @p keys is ignored.
 * If `keys == NULL`, the keyset will be empty with space reserved for at least @p keyCount keys.
 * Otherwise the keyset will contain all `keys[i]` for any `0 <= i < keyCount` where `keys[i] != NULL`.
 * In other words, `NULL`s within @p keys will be ignored and if @p keys contains more than @p keyCount keys
 * the additional ones will also be ignored.
 */
Key * elektraKeySetNew (Key ** keys, size_t keyCount);

> ```

Author: I decided to use a slightly more useful API than before. I still think that this API (especially `elektraKeyNew`) should almost never be used directly. However, the previous API was essentially just an allocator and had the problem that you always create a key with the wrong name first.

The fact that the size must be passed separately is a bit annoying, but since it should be hidden by another API in practice it doesn't matter much. The API should also still be simple enough that it could be implemented in any language.

Reviewer: Please write this in the decision (considered alternatives). It is unclear to me which text refers to Key and which to KeySet. In case of the Key** the size is actually not needed and it could be null-terminated.

> The API should also still be simple enough that it could be implemented in any language.

As said above, it was also possible to call keyNew from any language.

Other libraries will provide APIs on top of these functions.
These functions should be more idiomatic to their target language.

### Language Examples

> There are some examples of what the language specific APIs could look like in [another file](constructor_functions_examples.md).

Reviewer: I think you can also have these examples here (or in Notes)

Author: They are pretty long and don't mostly serve as an illustration during the discussion. I'm not sure it is worth keeping them at all, once we finalize the decision. It's only gonna lead to problems, if we later on decide on different APIs (very possible since the examples are not part of the actual decision here).

## Rationale

It is very hard to find a C API that works well in every language, but is still useful for creating keys.
It is much easier, to provide good APIs for a single language and build those on top of a very basic generic API.

This decision also guides Elektra towards the goal of a minimal core API.

> 

Reviewer: Argumentation for the actual API you chose is missing. As said: benchmarks would be highly appreciated.

Author: I will add the Considered Alternative (`keyNew(void)` and `ksNew(size_t)`) and a short explanation why other APIs would be worse (in particular for other languages).

> benchmarks would be highly appreciated.

What exactly would a benchmark achieve here? I didn't propose any specific optimizations, so I can't really benchmark anything.
